File: Python/Leaderboard/leaderboard.py
from typing import List, DefaultDict
from abc import ABC, abstractmethod
from collections import defaultdict


class AbstractDriver(ABC):

    @abstractmethod
    def get_name(self) -> str:
        pass


class Driver(AbstractDriver):
    def __init__(self, name: str, country: str):
        self.name = name
        self.country = country

    def get_name(self) -> str:
        return self.name


class SelfDrivingCar(AbstractDriver):
    def __init__(self, company: str, algorithm_version: str):
        self.company = company
        self.algorithm_version = algorithm_version

    def get_name(self):
        return "Self Driving Car - {} ({})".format(self.company, self.algorithm_version)


class AbstractRace(ABC):

    def __init__(self, name: str, results: List[AbstractDriver], points: List[int]):
        self.name = name
        self.results = results
        self._points = points

    @abstractmethod
    def points(self, driver) -> int:
        pass


class Race(AbstractRace):

    def __init__(self, name, results: List[AbstractDriver], points: List[int] = (25, 18, 15)):
        super().__init__(name, results, points)
        self.driver_names = [driver.get_name() for driver in self.results]

    def points(self, driver) -> int:
        return self._points[self.results.index(driver)] \
            if self._check_driver_in_points(driver) else 0

    def _check_driver_in_points(self, driver) -> bool:
        return driver in self.results and self.results.index(driver) < len(self._points)


class Leaderboard(object):

    def __init__(self, races: List[AbstractRace]):
        self.races = races
        self.total_driver_points = self.driver_points()

    def driver_points(self) -> DefaultDict:
        driver_points = defaultdict(int)
        for race in self.races:
            for driver in race.results:
                driver_points[driver.get_name()] += race.points(driver)
        return driver_points

    def driver_rankings(self) -> List:
        rankings = sorted(self.total_driver_points.items(), key=lambda x: x[1], reverse=True)
        return [name for (name, points) in rankings]

File: Ruby/LeaderBoard/leader_board.rb
class LeaderBoard
  def initialize(races)
    @races = races
  end

  def driver_points
    driver_points = {}
    @races.each do |race|
      race.results.each do |driver|
        name = race.driver_name(driver)
        driver_points[name] = driver_points.fetch(name, 0) + race.points(driver) 
      end
    end

    return driver_points
  end

  def driver_rankings
    rankings = driver_points.sort_by{|name, points| points}.reverse
    return rankings.collect{|name, points| name}
  end
end

class Driver
  attr_reader :name, :country
  def initialize(name, country)
    @name = name
    @country = country
  end
end

class SelfDrivingCar < Driver
  attr_accessor :algorithm_version
  def initialize(algorithm_version, company)
    super
    @algorithm_version = algorithm_version
    @company = company
  end
end

class Race
  @@points = [25, 18, 15]

  def initialize(name, drivers)
    @name = name
    @results = drivers
    @driver_names = {}
    @results.each do |driver|
      driver_name = driver.name
      if driver.class == SelfDrivingCar
        driver_name = "Self Driving Car - " + driver.country + " (" + driver.algorithm_version + ")";
      end
      @driver_names[driver] = driver_name
    end
  end

  def position(driver)
    @results.index driver
  end

  def points(driver)
    @@points[position(driver)]
  end

  def results
    @results
  end

  def driver_name(driver)
    @driver_names[driver]
  end    
end
